ls-ui-react-app/src/pages/UploadCircuitPage.tsx

```diff
@@ -51,26 +51,57 @@ const UploadACircuit = ( {appState, setAppState} : AppStateProps ) => {
         }).then( (response ) => {
-            setAppState({
-                ...appState,
-                request: true,
-                apiResponse: response,
-            })
-            try {
-                const responseJson = JSON.parse(response.data) as CompilationResult
+            
+            if (response.data.errorMessage) {
+                const errortype = response.data.errorType;
+                const msg = response.data.errorMessage;
                 setAppState({
-                    ...appState,
-                    compilationResult: responseJson,
+                    apiResponse: new CompilerError(errortype,msg),
+                    compilationIsLoading: false
                 })
             }
-            // catch errors due to parsing JSON
-            catch(err) {
-                // JSON parsing error
+            else{
+                try {
+                    const responseJson = JSON.parse(response.data) as CompilationResult;
+                    setAppState({
+                        apiResponse: new CompilationResultSuccess(responseJson.slices,responseJson.compilationText),
+                        compilationIsLoading: false
+                    })
+                } catch (error) {
+                    setAppState({
+                        apiResponse: new JsonParseError((error as Error).toString()),
+                        compilationIsLoading: false,
+                    })
+                }
             }
-        // Catch Axios Errors, ie API timeout
+
+        // Catch Axios Errors: API errors -> timeout, error compiling, or no response
         }).catch( (error) => {
-            console.error(error)
-            setAppState({...appState, errorMsg: error.toString(), compilationIsLoading: false, compilationResult:undefined, request:true ,apiResponse:undefined})
+
+            // needs more testing. First two requests after a long lambda cooldown period result in CORS / Network Errors
+            if (error.response) {
+                console.error("AXIOS error:")
+                console.log(error.response.data)
+                console.log(error.response.headers)
+                console.log(error.response.status)
+                const error_code = error.response.status;
+                const error_data = error.response.data;
+                const error_headers = error.response.headers;
+                // var api_error = new ApiHttpError(error_code,error_data,error_headers)
+                setAppState({
+                    // apiResponse: new ApiHttpError((error as Error).toString()),
+                    apiResponse: new ApiHttpError(error_code,error_data,error_headers),
+                    compilationIsLoading: false
+                })
+            }
+            else{
+                setAppState({
+                    // apiResponse: new ApiHttpError((error as Error).toString()),
+                    apiResponse: new NoServerResponse("Server did not respond"),
```

Comment:
Ok I see you set the string here

Reply:
You could leave the `NoServerResponse` class empty (using the type just as a tag) and give it a `getMsg` method to get this string, but it doesn't really matter

Reply:
> 

Challenge accepted!